### ISSUE TITLE
Consistent style, EditorConfig

### ISSUE THREAD

**Issue by iKarith** [linappleii/linapple#8]
We need a consistent style in the code.  @ghedger provided an example (two space indents) which #7 will mostly give us.  The project is at a crossroads though and since Greg hasn't joined the project yet we can ignore his preferences. 🙃

As long as what we've got is consistent (it isn't, and won't be after #7 is merged) I don't have any specific need for a particular style of indentation.  Except GNU braces are dumb, but nobody else uses them anyway. 😎

I am going to drop an [EditorConfig](http://editorconfig.org/) for whatever we go with in there.  That's trivial for vim-flavored editor users who use plugin managers and not much more complex if you don't. I might also add verbose vim modelines as hints for anyone not using an EditorConfig-supporting editor, mostly because I know how to write them.

We might want to look at whatever AppleWin uses most consistently and do that—that'd make it a lot easier to pull features from them.  They've had a dozen years' worth.

 to help do that in an automated fashion, and I propose to do that here as well.  Greg and I use vim flavors, which if not having EditorConfig built-in is at least a trivial addition if you use a plugin manager and only a little less so if you don't.  Once we have that in place, I don't especially care because I'll do what I normally do and the editor will do magic for me.  ;)

I'd propose though that we adopt the most consistent style I see in AppleWin's source code.  That's a four-space hard tab, which again as long as there's an EditorConfig file to configure, I don't mind.  I also typically add verbose vim modelines (since I don't know how to write them for other editors offhand) as hints for people who don't have a

**Comment by maxolasersquad** [comment]
Some analysis by looking at a handful of *.cpp files in the src directory. This is hardly thorough.

# Indentation
Two spaces

# Brackets

## Functions
GNU brackets

## Other
K&R

# Naming

## Class
No classes are defined.

## Functions
All over the place, but seems to mostly be PascalCase.

## Variables
camelCase

**Comment by maxolasersquad** [comment]
It's probably also a good idea to document what versions of C++ will be supported.

**Comment by iKarith** [comment]
Ah crud, you're right about needing to define C++ versions we support.  By default we're using C++98 since that's what gcc uses by default now.  I could do the UNIX thing where I sneer about required today, deprecated tomorrow, and forbidden in two weeks, but the fact is that I haven't a preference.  :)

LinApple appears to be written essentially in C.  It builds with a C++ compiler and uses the odd C++ template, which isn't a bad thing because C macros really are a dirty, ugly 40+ year old hack and C should've deprecated them 20 years ago with C99.

If we have reason to move to C++11 or C++14, I see no objection really.  If Linux is presently our primary platform, both gcc and clang support all three, and Apple and now Microsoft also use clang, so I think we're good to go there.

**Comment by iKarith** [comment]
I am in error.  Current versions of gcc use "gnu++14" as their default standard.  So gcc extensions supported (which we don't use) to c++14.  If we wish to maintain c++11 or c++98 compatibility, we should test that from time to time I suppose.

I think my current efforts for a port to SDL2 should probably be suspended for now.  Not because I couldn't do it eventually, but because the codebase is kind of too messy for that just now.

**Comment by maxolasersquad** [comment]
Putting an opinion out there, I think hard tabs are as face-palmy as gnu brackets, FWIW.

**Comment by iKarith** [comment]
You described the function brackets as being GNU above…  It's true that GNU does use opening brace on its own line, but it's also K&R, Allman, 1TBS, etc.  :)

The arguments against uncuddled braces are #ifdef and block readability.  I counter that with SDL we shouldn't have much #ifdef.  One could argue that this is what brace matching features of modern editors are for.  Even nano might support something like that.  A point of contention is stuff like:

```c++
    if (this)
        that();

    if (this) {
        that();
    }
```

My preference is the latter simply because people who write C compilers for a living seem to all have blog posts cautioning against the former and explaining how it absolutely will 100% of the time trip someone up in a project sooner or later, including themselves.

Alexey did a fair bit of:

```c++
    if (this) { that(); other(); thing(); }
```

…what is this, Perl?  I'll be inserting some newlines I think.

Spaces, tabs, it usually doesn't matter to me when there's an EditorConfig file.  What I want to avoid is 8 is the spacing of the tab key, and that tab key spacing shall be 8. to 81 columns shall thy lines not reach, nor shalt thou write functions longer than 25 lines.  Oh, and thou mayest not cuddle thy braces, for it is unseemly in polite company.  Have the people who write this s**t on stone tablets EVER tried to write a blit routine that has to deal with two different pixel formats, even without optimization?  I bet they have not.

By contrast I'd like to avoid Ivan's 800-line-long (not an exaggeration) conditional blocks. 🤪

When I _don't_ have an editorconfig file, my editor defaults to four-space hard tabs for C/C++ and 4 space soft tabs for Python, because for all the 8 space purists out there, 4 is what I encounter most often for C.  SDL is obnoxious as C goes—it uses, in vim parlance, tabstop=8 shiftwidth=4 noexpandtab.  C'mon Sam, who does that?!  Here's relevant bits of my old .vimrc:

```vim
   au BufNewFile,BufRead *.c,*.cpp,*.h setlocal ts=4 sw=4 noet                                   
   au BufNewFile,BufRead *.py setlocal ts=4 sw=4 et                                              
   au BufNewFile,BufRead *.sh,*.ac set ts=8 sw=8 noet                                            
```

Certainly all the trendy people say you use spaces now because tabs are SO 1990s and we all use web browsers to edit code now so we can't tab!  I'm not sympathetic. 😉  But I also am quite comfortable using either if the editorconfig is set up so I don't have to do it by hand.  Doing it the AppleWin way seems the best choice to me if only because I'd like to make LinApple less and less of a fork over time, rather than more and more.

**Comment by maxolasersquad** [comment]
I'm definitely on board the 8 spaces is crazy, and forces your code to hit that wall on the right way too quickly. I actually prefer 2 spaces because it really help avoiding the character limit per line, but most of my professional stuff is 4 and it's not so bad. Also I love Python and always write my Python with 4 spaces to be consistent with the language preferences.

I used to be fan of tabs, because it allows the programmer to configure their won white space, until someone showed me how it screws up indentations all over the place, and then you also don't know if you truly have or have not hit the character limit. It's easier to code 4 spaces, even if that's not your _preference_ than it is to deal with the inconsistencies it causes. I feel that vast majority has moved to this thinking as well.

I'm a fan, for the sake of consistencies, of all code blocks being wrapped in brackets. The following isn't so bad, but if my brain is expecting brackets and can create a mild bump trying to parse things quickly. This certainly isn't the kind of thing I push because I do understand why many people prefer it, and it's not so bad either way.

```c
    if (this)
        that();
```

My personal preferences for brackets is the example you gave.

```c
    if (this) {
        that();
    }
```

```c
    if (this) { that(); other(); thing(); }
```

That on is definitely bothersome to my brain. Again, trying to scan code it's difficult to understand there's a block here, and what's going on in that block.

In the end, consistency is more important when coding than any specific standard. Since I'm not likely to actually be contributing much to this project I would certainly yield to other's preferences. I just wanted to make sure my voice was at least part of the conversation.

**Comment by iKarith** [comment]
Draft for CONTRIBUTING.md written this morning:

> # Contributing to LinApple
> 
> If you'd like to contribute to LinApple, please feel free to submit pull
> requests via GitHub!  Members of the LinApple development team will likewise
> be using the PR mechanism for larger commits, although they may fix small bugs
> directly in-tree.
> 
> The justification for this development style are many.  First, it becomes
> easier for two people working on overlapping segments of code don't need to
> constantly deal with conflicts back and forth as individual commits are made.
> 
> **A word about coding style.**  LinApple is based somewhat loosely these days
> on AppleWin, which had a reasonably consistent style.  Developers since have
> had various inconsistent styles until one of our developers recently tried to
> make at least the indentation consistent.  A subsequent merge undid some of
> his work, and we're currently re-evaluating.  Please just try to match the
> existing style of code you're modifying for the time being.  Thank you for
> your patience, understanding, and willingness to even touch this mess if this
> transitional paragraph is still here for you to read. 😁
> 
> 
> ## Indentation
> 
> Most of the code uses a two-space indent at the moment.  Please use that and
> one statement per line for new code.  Don't worry about fixing surrounding old
> code at the moment.
> 
> 
> ## Brackets
> 
> ### Functions
> 
> Opening and closing braces of functions should begin on their own line.  Most
> C and C++ style guides use this design, AppleWin uses it, and 
> 
> ### Control statements
> 
> Cuddling braces on the same line as K&R/1TBS appears to be the norm.  Opening
> brace on the same line as the control statement.
> 
> 
> ## Naming
> 
> ### Classes
> 
> We have no classes.  If you have a good reason to change that, suggest
> something!
> 
> 
> ### Functions
> 
> Function naming is all over the place.  Mostly PascalCase, and currently
> without use of any kind of namespacing, be it using C++ or an informal
> C-style.
> 
> 
> ### Variables
> 
> A mix of camelCase variables and Windows-style "Hungarian notation". You
> should probably prefer camelCase.
> 
> 
> ## C++ Standard
> 
> LinApple is basically written in C, but compiled with a C++ compiler.  We
> build it with gcc on Linux, and recent versions of gcc are C++14 by default.
> It may built with LLVM-Clang for use on Mac and Windows as well, and Clang
> also supports C++14.

**Comment by maxolasersquad** [comment]
If you will pull the trigger on making an .editorconfig file I'll try to find some time this weekend to to reformatting the code. I feel I've made my opinions known. I'll accept whatever direction you decide to take.

**Comment by iKarith** [comment]
I've been holding off on this because I want to be not in the middle of something when I do that.  We're still in the midst of #8 which is the last technical thing left before I think we can consider my merge efforts to be complete.  (Well, complete enough—I still want macOS support, but that can wait for a bit.)  That just leaves this and I think I can push it out to Greg and Chris, and see what still needs to be done, if anything.

I think we could call the results of that all but a release, honestly.

**Comment by rhaleblian** [comment]
HI! I want to cherry-pick changes in @dabonetn 's copy of linapple back onto this repo, so was trying to do diffs, and needed to get some whitespace equivalence, and saw this. I'm running `clang-format` on each side, then diffing and manually merging. SO much better this way.
The files i saw at AppleWin (eg. Disk.cpp) use real tabs. Even if real tabs were not the ultimate destination, they may be a good neutral position to reach.

**Comment by rhaleblian** [comment]
Part of taking in the linapple-pie work
https://github.com/yoyodyne-research/linapple/tree/retropie
included converting the source on both sides using clang-format, so that i could hand-merge two repos without shared history.

**Comment by maxolasersquad** [comment]
This is pretty close to the style of formatting that is my preference. FWIW.

**Comment by rhaleblian** [comment]
Cool. Great hair.

**Comment by ghedger** [comment]
Looks good.  Looking at just a couple of the files, I like (and prefer) the 2-space indent and brackets on same line.  Someone had to do it, so thanks.

**Comment by rhaleblian** [comment]
Sure, i recently started just letting the machine handle formatting via formatters and i've found it frees my mind to stay focused on the core problem at hand. We rely on compilers to perform the repetitive task of formatting the code into machine language, so why not let it apply consistency rules to source, being rules we can describe to machine, right?

**Comment by rhaleblian** [comment]
ps. one thing i think happened during the merge is that @dabonetn 's implementation of disk image command line options trumped the ones here. I guess they are functionally identical; the former used '-1' where here you guys used '-d1', think?

**Comment by ghedger** [comment]
-d1 seems more logical and consistent with command line options in other programs.

**Comment by ghedger** [comment]
I have a question for all: Do we want to adopt a standard procedure for keeping the linapple changes in a common repo?  I thought linappleii/linapple had become the "official" one.  I have my own repo (ghedger/linapple).  The reason for the question:  I had issued several pull requests last Fall and Winter for changes.  These changes were never pulled into linappleii/linapple, wjhich I think would be the norm -- although I'm unaccustomed to usual practice for multiple-developed projects on github, normally just using it as a SCM for my own projects.

My apologies if this is an overly basic and elementary question.

**Comment by maxolasersquad** [comment]
@ghedger I think it would bring sanity to the project to have a single repository that is considered _official_, like most projects do.

I think all this version of it needs is someone to take up the mantle, even if that just means reviewing pull requests and accepting them. I'm doing what I can on that front, but my skills are in web development and business processing, not emulation, so I'm not really the best to judge the quality of PRs or understand the details of what they do.

**Comment by rhaleblian** [comment]
> a single repository that is considered official

Yep yep. I'm considering this repo the official one, short of applewin itself. I didn't want the RetroPie changes to languish while upstream improves. Eventually the former will be imcompatible with the latter.

> keeping the linapple changes in a common repo

A team could work in feature branches in upstream, or submit PRs to master from a fork, or push to an upstream feature branch from a fork and PR from there. I prefer to work in feature branches. If i'm doing that in a fork:

1. The master in my fork is always identical to upstream so i can always see how compatible the feature work is and pre-integrate it to avoid merge conflicts during the PR. I can always pull master into my master cleanly.

2. When i push a branch upstream, my feature is known to upstream without needing to know about any other forks. If my fork disappeared, the work is not lost. My fork is, in fact, meant to be ephemeral; I don't want to fork and rebrand this project, i would prefer it support RetroPie intrinsically. Gracious, Github has enough forks already.

Since i am using a branch, i could also do the work directly into upstream, maybe i don't need a fork.

Ultimately though, yes -- shareable features and fixes get back to this repo somehow.

**Comment by rhaleblian** [comment]
> if that just means reviewing pull requests and accepting them

i might be able to do some, but i don't want to force my ideas about what should go into the base application. if a PR is already understood to be desired, then that's less of a problem.

i might not know what a given PR might break. but on the other hand, it's a way to learn the code.

**Comment by ghedger** [comment]
The basis for the question is this:  The last time I tried to contribute to an "official" linapple repository, the owner got sick!  PRs were ignored, and I had time to work on LinApple, so I just, well, kept working in my own repo, still issuing the PRs, but to share the latest work with anyone meant pointing them to my repo.

It seems every couple years someone new takes ownership of LinApple and then forgets about it.

Other topic: There is an ongoing bug in AppleWin, and thus LinApple, relating to bank switching in High RAM -- these are the $C08x switches.  AppleWin is (mis)using the AltZP as a switch for this in a long, complex logic statement.  It's been a few months or I'd be more specific.  Anyways, it's a case where it "works" for most software, since using the two upper banks of LCRAM was apparently not common, but is nevertheless something I'd like to take a look at when I'm not embroiled in paying projects.

When that is fixed (when I can have a place to set up my real Apple IIe), it will represent a divergence from AppleWin.

**Comment by maxolasersquad** [comment]
I think we just need to build a group of "core" developers; people willing to take ownership of this and work together to move the project forward. It looks like @iKarith won't be able to work on this like he was hoping to, at least not in the near future. Anyone who is willing or has thoughts on this, please continue that discussion in #42.

**Comment by ghedger** [comment]
Agreed, and I do hope the best for @iKarith in his recovery.

I might suggest this core team be large enough that there's a high probability someone can service the PRs (we all get busy after all).  I would also be happy working on the feature branch / PR model as opposed to a fork provided the permissions are there.

> i might not know what a given PR might break. but on the other hand, it's a way to learn the code.

It's all part of the learning process -- as long as new dependency libs, build flags, etc. are documented, should be okay (i.e. no one tries to update to the latest SDL without some serious discussion first :) ).  I don't want to say too much as I wasn't available last year at this time during the last heavy spate of development whereas many of you were...

**Comment by iKarith** [comment]
My intention was to add those who've done major work to it to the project as developers and a couple of others as owners so that there was no single point of failure. I've had a lot of medical issues crop up and other projects have to be a higher priority for now—but if there are multiple developers who have access, whoever can provide oversight at any given moment can tend to pending things.

I figured it still makes sense to work on things in separate branches or repos (however makes sense) and PR them for comments before stuff gets merged in. That way if there's something that is proposed that doesn't work for some reason, there's no need for "commit terrorism" of patch/revert cycles.

I did intend to take a much more active role in this project right away. As I'm not doing that anytime really soon, I can at least describe what I had in mind and see what y'all think. Someone might want to pick up where I left off:

 I wanted to change the config file to be a "more traditional INI format" (as it's currently just kinda sloppy quasi-INI), even if that means that people need to redo their config files. In fact I assumed that'd be necessary, but it shouldn't be too bad since a script ought to be able to read in one and write out the other.

A big priority was to actually have properly UNIXy/multi-user-safe filesystem expectations: Reading defaults in in a sysconfdir, getting ROMs, fonts, etc. from eg. a sharedir/rom and disk images from a sharedir/disks and possibly tape images from a sharedir/tapes, all of which could be overridden at build time and have reasonable defaults on POSIX-like systems and Windows. The idea of a Mac bundle is in the back of my head for the future, but wasn't a priority. I figured on the user side, you'd have essentially an XDG sort of layout Linux, essentially ~/.config/linapple/linapple.ini for overriding syscofdir/linapple.ini and for specifying where to look for disk images, etc.

(I suspect the LinApple default for system/shared data would be /usr/share/linapple/{roms,disks,…} but on a Raspple II system, I'd probably change the default to put them in /usr/share/apple2/disks or something to allow images to be shared with other emulators, but I don't think that makes a great default and again it's an easy "distribution choice" thing to override.)

I also wanted to begin looking at the question of implementing .woz support at some point, and generally try and clean up the codebase where that makes some sense. Mostly that is a low priority for me because it requires a deeper understanding of the codebase than I have and a lot of time spent with John's excellent spec for the format. No doubt the spec will be quite explicit what to do and how to do it, but I'm not quite familiar enough to know what those look like in LinApple terms.

"Clean up the codebase"  … was supposed to help me accomplish the former, and generally a lot of other things. Now the problem there is that there's multiple ways to clean it up. One would be to follow the AppleWin code as closely as possible, which makes borrowing features easier. Advantage there is that if we ever get feature parity, the possibility of two projects becoming one that just works cross-platform does exist.

I mean, I don't have much hope for it, but it'd be possible. That's why my thought was the other direction made more sense: Don't bother to try to keep the code close to AppleWin, as even though it has improved, the code isn't very well delineated into separate modules with clear interfaces. I don't feel the need to go and C++ everything with OOP-or-die strategies mentality, but having things init in a structured way and then shutdown essentially in reverse order, with the program operating on a non-blocking loop in between intended to make machine cycles easier… that seems a logical plan to me. (There's some blocking code in LinApple, particularly features added that weren't in AppleWin at the time…)

There's other peripherals we could emulate, AppleWin features we could implement, etc. But I'd have been just happy doing what I've described here and consider that to have done all that really actually needed doing. Anything else would be just cool stuff I also got a chance to play with.

**Comment by rhaleblian** [comment]
My all-in reformatting left me in a bad place re integrating new PR merges, as they are on old formatting, so i'm backing off on what i've done, and redo the backports in pieces.

> "Clean up the codebase" … was supposed to help me accomplish the former, and generally a lot of other things. Now the problem there is that there's multiple ways to clean it up. One would be to follow the AppleWin code as closely as possible, which makes borrowing features easier. Advantage there is that if we ever get feature parity, the possibility of two projects becoming one that just works cross-platform does exist.

I think so too (despite my talk about libretro; upstream might be easier to port to libretro, actually). I was looking at bits at "the edges" that could achieve parity, no conclusions yet.

Upstream is using tabs. If you set your tabs to 2 spaces, if would be a lot like the formatting i did.

**Comment by ghedger** [comment]
Hello Ray, it's a complex labyrinth, yes.  Am following progress.  Not able to contribute at the moment (living in CAD the past several weeks for a completely unrelated project), but expect to in the future.

My vote would be for eliminating tabs altogether in favor of spaces (ideally, two spaces), but I'm not staring at the code at the moment like you are, so it's in your capable hands.

**Comment by maxolasersquad** [comment]
I used this conversation as the basis for creating PR https://github.com/linappleii/linapple/pull/64 with an editorconfig file and formatting all of the code. I've created it as a draft PR so as to allow a conversation regarding the choices here before/if it gets pulled in.
